### ISSUE TITLE
Rollback of committed entries

### ISSUE THREAD

**Issue by alesapin** [eBay/NuRaft#185]
Hi! I'm testing my ZooKeeper-like system based on NuRaft using Jepsen framework. Recently I found a very strange behavior when NuRaft decided to rollback already committed log entries:
```
2021.03.24 11:23:01.057713 [ 388 ] {} <Information> RaftInstance: rollback logs: 509 - 530, commit idx req 400, quick 530, sm 530, num log entries 1, current count 0                                                                          
2021.03.24 11:23:01.057726 [ 388 ] {} <Warning> RaftInstance: rollback quick commit index from 530 to 508                                                                                                                                      
2021.03.24 11:23:01.057740 [ 388 ] {} <Warning> RaftInstance: rollback sm commit index from 530 to 508                                                                                                                                  
2021.03.24 11:23:01.057751 [ 388 ] {} <Information> RaftInstance: rollback log 530                                                                                                                                                             
2021.03.24 11:23:01.057766 [ 388 ] {} <Information> RaftInstance: rollback log 529                                                                                                                                                             
2021.03.24 11:23:01.057780 [ 388 ] {} <Information> RaftInstance: rollback log 528                                                                                                                                                             
2021.03.24 11:23:01.057788 [ 388 ] {} <Information> RaftInstance: rollback log 527                                                                                                                                                             
2021.03.24 11:23:01.057794 [ 388 ] {} <Information> RaftInstance: rollback log 526                                                                                                                                                             
2021.03.24 11:23:01.057801 [ 388 ] {} <Information> RaftInstance: rollback log 525                                                                                                                                                             
2021.03.24 11:23:01.057807 [ 388 ] {} <Information> RaftInstance: rollback log 524                                                                                                                                                             
2021.03.24 11:23:01.057813 [ 388 ] {} <Information> RaftInstance: rollback log 523                                                                                                                                                             
2021.03.24 11:23:01.057828 [ 388 ] {} <Information> RaftInstance: rollback log 522
2021.03.24 11:23:01.057860 [ 388 ] {} <Information> RaftInstance: rollback log 521
2021.03.24 11:23:01.057867 [ 388 ] {} <Information> RaftInstance: rollback log 520
2021.03.24 11:23:01.057873 [ 388 ] {} <Information> RaftInstance: rollback log 519
2021.03.24 11:23:01.057879 [ 388 ] {} <Information> RaftInstance: rollback log 518
2021.03.24 11:23:01.057885 [ 388 ] {} <Information> RaftInstance: rollback log 517
2021.03.24 11:23:01.057891 [ 388 ] {} <Information> RaftInstance: rollback log 516
2021.03.24 11:23:01.057897 [ 388 ] {} <Information> RaftInstance: rollback log 515
2021.03.24 11:23:01.057903 [ 388 ] {} <Information> RaftInstance: rollback log 514
2021.03.24 11:23:01.057909 [ 388 ] {} <Information> RaftInstance: rollback log 513 
2021.03.24 11:23:01.057915 [ 388 ] {} <Information> RaftInstance: rollback log 512
2021.03.24 11:23:01.057920 [ 388 ] {} <Information> RaftInstance: rollback log 511
2021.03.24 11:23:01.057926 [ 388 ] {} <Information> RaftInstance: rollback log 510
2021.03.24 11:23:01.057931 [ 388 ] {} <Information> RaftInstance: rollback log 509
2021.03.24 11:23:01.057937 [ 388 ] {} <Information> RaftInstance: overwrite at 509
2021.03.24 11:23:01.209754 [ 388 ] {} <Information> RaftInstance: receive a config change from leader at 509
2021.03.24 11:23:01.209816 [ 388 ] {} <Debug> RaftInstance: [after OVWR] log_idx: 510, count: 1
```

As far as I understand this should never happen in RAFT and the comment clearly states about it https://github.com/ebay/NuRaft/blob/master/src/handle_append_entries.cxx#L631-L645. So isn't it a bug? 

Also, I've found this interesting comment:
https://github.com/ebay/NuRaft/blob/master/src/handle_append_entries.cxx#L696-L717.

Full trace logs from all three nodes:
1) Node which rolled back: https://gist.github.com/alesapin/5bcd85abd390f723cc30e0a63f0abd44
2) Other node1: https://gist.github.com/alesapin/f6d8754f7a2f749e376c8a0522bf6c43
3) Other node2: https://gist.github.com/alesapin/fa0e5af16152415f93942ed77bb7819a

Related https://github.com/ClickHouse/ClickHouse/pull/21677

**Comment by alesapin** [comment]
Also it seems like it leads to actual data loss, but I'm not 100% sure:
```
INFO [2021-03-23 17:36:27,930] jepsen test runner - jepsen.core {:perf {:latency-graph {:valid? true},
        :rate-graph {:valid? true},
        :valid? true},
 :workload {:ok-count 156,
            :valid? false,
            :lost-count 3,
            :lost "#{231 233 242}",
            :acknowledged-count 159,
            :recovered "#{}",
            :ok "#{0..1 4 8 12 14 18..20 23..24 30 32 42 45..46 48 55 60 67 72 77 80 82 84 86..87 89 97..98 101 106 109 111 118 122..124 126 128 130 132 134 136 138 142 146 148..149 151 153..154 156 159 163..164 167 171 173 176 178..179 182 185..186 188 191 193 196 198 201..205 209..214 216..217 219 222..223 227 229 287 305 308 345 358 362 369 375 378 386 396 405 417 425 434 445 454 462 470 480 491 498 505 516 526 534 542 552 563 571 578 587 598 607 615 670..671 674 685 704 718 734 749 765 785 807 821 841 858 874 887 907 929 943 963 979 1000 1024..1030 1032 1035 1037 1039}",
            :attempt-count 1041,
            :unexpected "#{}",
            :unexpected-count 0,
            :recovered-count 0},
 :valid? false}


Analysis invalid! (ﾉಥ益ಥ）ﾉ ┻━┻
```

**Comment by alesapin** [comment]
Detailed scenario:
1. Node3 (man) commit entries to quorum up to log idx = 530
```
2021.03.24 11:22:51.547034 [ 388 ] {} <Debug> RaftInstance: Response back a append_entries_response message to 2 with Accepted=1, Term=1, NextIndex=531
2021.03.24 11:22:51.547049 [ 391 ] {} <Trace> RaftInstance: commit_cv_ wake up
2021.03.24 11:22:51.547113 [ 391 ] {} <Debug> RaftInstance: commit upto 530, current idx 528
2021.03.24 11:22:51.547138 [ 391 ] {} <Trace> RaftInstance: commit upto 530, current idx 529
2021.03.24 11:22:51.547183 [ 391 ] {} <Trace> RaftInstance: commit upto 530, current idx 530
2021.03.24 11:22:51.547216 [ 391 ] {} <Debug> RaftInstance: DONE: commit upto 530, current idx 530
```
2. Node2 (sas) confirms and also commits them up to 530
```
2021.03.24 11:22:51.536502 [ 1096 ] {} <Debug> RaftInstance: commit upto 530, current idx 528
2021.03.24 11:22:51.536513 [ 1096 ] {} <Trace> RaftInstance: commit upto 530, current idx 529
2021.03.24 11:22:51.536561 [ 1096 ] {} <Debug> RaftInstance: notify cb 529 0x7faa1bb4f320
2021.03.24 11:22:51.536579 [ 1096 ] {} <Trace> RaftInstance: commit upto 530, current idx 530
```

3.  Node2 killed. Fresh logs and snapshots became corrupted:
```
2021.03.24 11:22:54.927817 [ 1259 ] {} <Warning> NuKeeperStateMachine: Failed to load from snapshot with index 500, with error DB::Exception: Checksum doesn't match: corrupted data. Reference: 61c9d2372f6e6d9e158c43cd0f3cefe1. Actual: 973c
4c18e43ee372e9bf60948bd6925e. Size of compressed block: 701, will remove it from disk
2021.03.24 11:22:54.927966 [ 1259 ] {} <Debug> NuKeeperStateMachine: Trying to load state machine from snapshot up to log index 400
2021.03.24 11:22:54.928080 [ 1259 ] {} <Debug> NuKeeperStateMachine: Loaded snapshot with last committed log index 400
2021.03.24 11:22:54.935879 [ 1259 ] {} <Warning> NuKeeperLogStore: Cannot completely read changelog on path /home/robot-clickhouse/db/coordination/logs/changelog_1_10000.bin, error: Checksums doesn't match for log /home/robot-clickhouse/db
/coordination/logs/changelog_1_10000.bin (version 0), index 509, blob_size 321
2021.03.24 11:22:54.935918 [ 1259 ] {} <Trace> NuKeeperLogStore: Totally read from changelog /home/robot-clickhouse/db/coordination/logs/changelog_1_10000.bin 508 entries
2021.03.24 11:22:54.935934 [ 1259 ] {} <Trace> NuKeeperLogStore: Continue to write into /home/robot-clickhouse/db/coordination/logs/changelog_1_10000.bin
2021.03.24 11:22:54.935953 [ 1259 ] {} <Warning> NuKeeperLogStore: Read finished with error, truncating all broken log entries
2021.03.24 11:22:54.972711 [ 1259 ] {} <Information> RaftInstance:    === INIT RAFT SERVER ===
commit index 400
term 0
election timer allowed
log store start 391, end 508
config log idx 0, prev log idx 0
2021.03.24 11:22:54.972867 [ 1259 ] {} <Information> RaftInstance: peer 1: DC ID 0, ch-iva.iva.yp-c.yandex.net:9444, voting member, 1
peer 2: DC ID 0, ch-sas.sas.yp-c.yandex.net:9444, voting member, 1
peer 3: DC ID 0, ch-man.man.yp-c.yandex.net:9444, voting member, 1
my id: 2, voting_member
num peers: 2
```
It restores at 508 entry.

4. Node2 participate in leader election increasing it's term

```
2021.03.24 11:22:56.574747 [ 1268 ] {} <Debug> RaftInstance: Receive a pre_vote_request message from 1 with LastLogIndex=501, LastLogTerm=1, EntriesLength=0, CommitIndex=501 and Term=3
2021.03.24 11:22:56.574844 [ 1268 ] {} <Information> RaftInstance: [PRE-VOTE REQ] my role follower, from peer 1, log term: req 1 / mine 1
last idx: req 501 / mine 508, term: req 3 / mine 3
HB dead
2021.03.24 11:22:56.574885 [ 1268 ] {} <Information> RaftInstance: pre-vote decision: O (grant)
2021.03.24 11:22:56.574897 [ 1268 ] {} <Debug> RaftInstance: Response back a pre_vote_response message to 1 with Accepted=1, Term=3, NextIndex=509
2021.03.24 11:22:58.103786 [ 1269 ] {} <Debug> RaftInstance: Receive a pre_vote_request message from 1 with LastLogIndex=501, LastLogTerm=1, EntriesLength=0, CommitIndex=501 and Term=4
2021.03.24 11:22:58.103846 [ 1269 ] {} <Information> RaftInstance: [PRE-VOTE REQ] my role follower, from peer 1, log term: req 1 / mine 1
last idx: req 501 / mine 508, term: req 4 / mine 3
HB dead
2021.03.24 11:22:58.103858 [ 1269 ] {} <Information> RaftInstance: pre-vote decision: O (grant)
2021.03.24 11:22:58.103868 [ 1269 ] {} <Debug> RaftInstance: Response back a pre_vote_response message to 1 with Accepted=1, Term=4, NextIndex=509
2021.03.24 11:22:58.111578 [ 1270 ] {} <Debug> RaftInstance: Receive a request_vote_request message from 1 with LastLogIndex=501, LastLogTerm=1, EntriesLength=0, CommitIndex=501 and Term=5
2021.03.24 11:22:58.111613 [ 1270 ] {} <Trace> RaftInstance:   FOLLOWER
2021.03.24 11:22:58.111680 [ 1270 ] {} <Information> RaftInstance: [PRIORITY] decay, target 1 -> 1, mine 1
2021.03.24 11:22:58.111697 [ 1270 ] {} <Information> RaftInstance: [VOTE REQ] my role follower, from peer 1, log term: req 1 / mine 1
last idx: req 501 / mine 508, term: req 5 / mine 5
priority: target 1 / mine 1, voted_for -1
2021.03.24 11:22:59.759932 [ 1269 ] {} <Debug> RaftInstance: Receive a request_vote_request message from 1 with LastLogIndex=501, LastLogTerm=1, EntriesLength=0, CommitIndex=501 and Term=6
2021.03.24 11:22:59.759990 [ 1269 ] {} <Trace> RaftInstance:   FOLLOWER
2021.03.24 11:22:59.760001 [ 1269 ] {} <Trace> RaftInstance: cancel existing timer
2021.03.24 11:22:59.760019 [ 1269 ] {} <Trace> RaftInstance: re-schedule election timer
2021.03.24 11:22:59.760038 [ 1269 ] {} <Information> RaftInstance: [PRIORITY] decay, target 1 -> 1, mine 1
2021.03.24 11:22:59.760051 [ 1269 ] {} <Information> RaftInstance: [VOTE REQ] my role follower, from peer 1, log term: req 1 / mine 1
last idx: req 501 / mine 508, term: req 6 / mine 6
priority: target 1 / mine 1, voted_for -1
```
At some point term increased to 7 and this node became leader:
```
2021.03.24 11:23:00.949721 [ 1268 ] {} <Debug> RaftInstance: Receive a request_vote_response message from peer 1 with Result=1, Term=7, NextIndex=502
2021.03.24 11:23:00.949730 [ 1268 ] {} <Trace> RaftInstance: src: 1, dst: 2, resp->get_term(): 7
2021.03.24 11:23:00.949743 [ 1268 ] {} <Information> RaftInstance: [VOTE RESP] peer 1 (O), resp term 7, my role candidate, granted 2, responded 2, num voting members 3, quorum 2
2021.03.24 11:23:00.949751 [ 1268 ] {} <Information> RaftInstance: Server is elected as leader for term 7
2021.03.24 11:23:00.949770 [ 1268 ] {} <Information> RaftInstance: number of pending commit elements: 0
2021.03.24 11:23:00.949788 [ 1268 ] {} <Information> RaftInstance: state machine commit index 400, precommit index 508, last log index 508
2021.03.24 11:23:00.949814 [ 1268 ] {} <Trace> RaftInstance: peer 3, interval: 500
2021.03.24 11:23:00.949828 [ 1268 ] {} <Trace> RaftInstance: peer 1, interval: 500
2021.03.24 11:23:00.949860 [ 1268 ] {} <Information> RaftInstance: [BECOME LEADER] appended new config at 509
2021.03.24 11:23:00.956001 [ 1271 ] {} <Trace> RaftInstance: resp of req 2 -> 3, type pre_vote_request, OK
2021.03.24 11:23:01.045067 [ 1268 ] {} <Trace> RaftInstance: send request to 3
2021.03.24 11:23:01.045151 [ 1268 ] {} <Trace> RaftInstance: last_log_idx: 508, starting_idx: 391, cur_nxt_idx: 510
2021.03.24 11:23:01.045167 [ 1268 ] {} <Debug> RaftInstance: append_entries for 3 with LastLogIndex=508, LastLogTerm=1, EntriesLength=1, CommitIndex=400, Term=7, peer_last_sent_idx 0
2021.03.24 11:23:01.045176 [ 1268 ] {} <Debug> RaftInstance: idx: 509
2021.03.24 11:23:01.045187 [ 1268 ] {} <Trace> RaftInstance: send req 2 -> 3, type append_entries_request
2021.03.24 11:23:01.045269 [ 1268 ] {} <Trace> RaftInstance: sent
2021.03.24 11:23:01.045283 [ 1268 ] {} <Trace> RaftInstance: send request to 1
2021.03.24 11:23:01.045291 [ 1268 ] {} <Trace> RaftInstance: last_log_idx: 508, starting_idx: 391, cur_nxt_idx: 510
2021.03.24 11:23:01.045302 [ 1268 ] {} <Debug> RaftInstance: append_entries for 1 with LastLogIndex=508, LastLogTerm=1, EntriesLength=1, CommitIndex=400, Term=7, peer_last_sent_idx 0
2021.03.24 11:23:01.045309 [ 1268 ] {} <Debug> RaftInstance: idx: 509
2021.03.24 11:23:01.045319 [ 1268 ] {} <Trace> RaftInstance: send req 2 -> 1, type append_entries_request
2021.03.24 11:23:01.045356 [ 1268 ] {} <Trace> RaftInstance: sent
2021.03.24 11:23:01.045369 [ 1268 ] {} <Information> RaftInstance:   === LEADER (term 7) ===
```
So for some reason, Node1 (iva) voted for Node2 (sas) in term 7, but this node doesn't have all the required log entries (log index 508 vs 530 on Node3).

After that Node2 received request to vote from Node3, but it was denied:
```
2021.03.24 11:23:01.045446 [ 1271 ] {} <Debug> RaftInstance: Receive a pre_vote_response message from peer 3 with Result=1, Term=6, NextIndex=531
2021.03.24 11:23:01.045493 [ 1271 ] {} <Trace> RaftInstance: src: 3, dst: 2, resp->get_term(): 6
2021.03.24 11:23:01.045509 [ 1271 ] {} <Information> RaftInstance: [PRE-VOTE RESP] peer 3 (O), term 6, resp term 6, my role leader, dead 3, live 0, num voting members 3, quorum 2
2021.03.24 11:23:01.045517 [ 1271 ] {} <Information> RaftInstance: [PRE-VOTE DONE] SUCCESS, term 6
2021.03.24 11:23:01.045526 [ 1271 ] {} <Information> RaftInstance: [PRE-VOTE DONE] actual vote is already initiated, do nothing
```

So the problem happened because Node1 denied to vote for Node3:
```
2021.03.24 11:22:59.978230 [ 4031 ] {} <Debug> RaftInstance: Receive a request_vote_request message from 3 with LastLogIndex=530, LastLogTerm=1, EntriesLength=0, CommitIndex=530 and Term=6
2021.03.24 11:22:59.978302 [ 4031 ] {} <Information> RaftInstance: [VOTE REQ] my role candidate, from peer 3, log term: req 1 / mine 1
last idx: req 530 / mine 501, term: req 6 / mine 6
priority: target 1 / mine 1, voted_for 1
2021.03.24 11:22:59.978331 [ 4031 ] {} <Information> RaftInstance: decision: X (deny), term 6
```
But voted for stale Node2:
```
2021.03.24 11:23:00.944439 [ 4033 ] {} <Information> RaftInstance: [VOTE REQ] my role follower, from peer 2, log term: req 1 / mine 1
last idx: req 508 / mine 501, term: req 7 / mine 7
priority: target 1 / mine 1, voted_for -1
2021.03.24 11:23:00.944454 [ 4033 ] {} <Information> RaftInstance: decision: O (grant), voted_for 2, term 7
2021.03.24 11:23:00.944466 [ 4033 ] {} <Debug> RaftInstance: Response back a request_vote_response message to 2 with Accepted=1, Term=7, NextIndex=502
```

**Comment by alesapin** [comment]
Seems like **Leader Completeness** was broken?
```
Leader Completeness: if a log entry is committed in a given term, then that entry will be present in the logs of the leaders for all higher-numbered terms.
```

**Comment by greensky00** [comment]
Hi @alesapin 

The basic foundation of Raft is that once a log is committed, any possible combinations of quorum should have at least one node that has the committed log. That means, at least one node in `{Node1, Node2}` or `{Node2, Node3}` or `{Node1, Node3}` should have the log 530.

But in the case you shared, looks like `{Node1, Node2}` was the quorum at the time the log 530 was committed, thus they have the log 530. But after `Node2` restarted, it lost the log 530 which should not happen. That causes the situation that the quorum set `{Node2, Node3}` doesn't include any node that has the committed log 530, and it ends up with such an undesired rollback.

So in summary, nodes should not lose the Raft log they received even after restart. Or, in other words, at least one node in a quorum set should have the last committed log.

**Comment by greensky00** [comment]
+) Looks like your question is about "why Node1 denied Node3?", the reason why is Node1 already voted for itself (Node1) for the term 6. A node cannot vote for multiple nodes in the same round of election:
```
2021.03.24 11:22:59.978302 [ 4031 ] {} <Information> RaftInstance: [VOTE REQ] my role candidate, from peer 3, log term: req 1 / mine 1
last idx: req 530 / mine 501, term: req 6 / mine 6
priority: target 1 / mine 1, voted_for 1
```
There is a log `voted_for 1`, which means that it requested a vote for other nodes for the term 6, at the same time voted for itself.  This is an expected behavior. It voted for Node2 at the different round, that is term 7.

**Comment by alesapin** [comment]
Yes, thanks for the explanation. I've read the RAFT article one more time and seems like we just break one of the requirements for the **Leader completeness** property, i.e:

```
2. leader_T replicated the entry on a majority of the cluster, and leader_U received votes from a majority of the cluster. Thus, at least one server (“the voter”) both accepted the entry from leader_T and voted for leader_U, as shown in Figure 9. The voter is key to reaching a contradiction.
```

In my case, this statement is not true:
`Thus, at least one server (“the voter”) both accepted the entry from leader_T and voted for leader_U`
We don't have such server because of disk corruption, so **Leader Completeness** was broken.

But rollback of SM seems like an awful case when we probably lose data. I think we should increase the log level to at least `error` in this case. Otherwise, it looks like a minor issue in logs.

**Comment by greensky00** [comment]
Makes sense, thanks for your PRs.
